agent/stats/engine.go

```diff
@@ -0,0 +1,446 @@
+	if err != nil {
+		log.Debug("Could not map container to task, ignoring", "err", err, "id", dockerID)
+		return
+	}
+
+	_, taskExists := engine.tasksToContainers[task.Arn]
+	if !taskExists {
+		log.Debug("Container not being watched", "id", dockerID)
+		return
+	}
+
+	// task arn exists in map.
+	container, containerExists := engine.tasksToContainers[task.Arn][dockerID]
+	if !containerExists {
+		// container arn does not exist in map.
+		log.Debug("Container not being watched", "id", dockerID)
+		return
+	}
+
+	container.StopStatsCron()
+	delete(engine.tasksToContainers[task.Arn], dockerID)
+	log.Debug("Deleted container from tasks", "id", dockerID)
+
+	if len(engine.tasksToContainers[task.Arn]) == 0 {
+		// No containers in task, delete task arn from map.
+		delete(engine.tasksToContainers, task.Arn)
+		// No need to verify if the key exists in tasksToDefinitions.
+		// Delete will do nothing if the specified key doesn't exist.
+		delete(engine.tasksToDefinitions, task.Arn)
+		log.Debug("Deleted task from tasks", "arn", task.Arn)
+	}
+}
+
+// GetInstanceMetrics gets all task metrics and instance metadata from stats engine.
+func (engine *DockerStatsEngine) GetInstanceMetrics() (*ecstcs.MetricsMetadata, []*ecstcs.TaskMetric, error) {
+	var taskMetrics []*ecstcs.TaskMetric
+	idle := engine.isIdle()
+	engine.metricsMetadata.Idle = &idle
+	if idle {
+		log.Debug("Instance is idle. No task metrics to report")
+		return engine.metricsMetadata, taskMetrics, nil
+	}
+
+	for taskArn := range engine.tasksToContainers {
+		containerMetrics, err := engine.getContainerMetricsForTask(taskArn)
+		if err != nil {
+			log.Debug("Error getting container metrics for task", "err", err, "task", taskArn)
+			continue
+		}
+
+		if len(containerMetrics) == 0 {
+			log.Debug("Empty containerMetrics for task, ignoring", "task", taskArn)
+			continue
+		}
+
+		taskDef, exists := engine.tasksToDefinitions[taskArn]
+		if !exists {
+			log.Debug("Could not map task to definition", "task", taskArn)
+			continue
+		}
+
+		taskMetrics = append(taskMetrics, &ecstcs.TaskMetric{
+			TaskArn:               &taskArn,
+			TaskDefinitionFamily:  &taskDef.family,
+			TaskDefinitionVersion: &taskDef.version,
+			ContainerMetrics:      containerMetrics,
+		})
+	}
+
+	if len(taskMetrics) == 0 {
+		// Not idle. Expect taskMetrics to be there.
+		return nil, nil, errors.New("No task metrics to report")
+	}
+
+	return engine.metricsMetadata, taskMetrics, nil
+}
+
+func (engine *DockerStatsEngine) isIdle() bool {
+	return len(engine.tasksToContainers) == 0
+}
+
+// initDockerClient initializes engine's docker client.
+func (engine *DockerStatsEngine) initDockerClient() error {
+	if engine.client == nil {
+		client, err := ecsengine.NewDockerGoClient()
+		if err != nil {
+			return err
+		}
+		engine.client = client
+	}
+
+	return nil
+}
+
+// addExistingContainers lists existing containers and adds them to the engine.
+func (engine *DockerStatsEngine) addExistingContainers() error {
```

Comment:
Can you move this up below `listContainersAndStartEventHandler`, which calls it?